crates/lib3h/src/transport/memory_mock/transport_memory.rs

```diff
@@ -233,53 +271,75 @@ impl Transport for TransportMemory {
+                    if let TransportEvent::IncomingConnectionEstablished(in_cid) = event {
+                        let to_connect_uri = my_server
+                            .get_inbound_uri(&in_cid)
+                            .expect("Should allways have uri");
+                        to_connect_list.push((to_connect_uri.clone(), in_cid.clone()));
                     } else {
-                        outbox.push(event.clone());
+                        output.push(event);
                     }
                 }
             }
         }
         // Connect back to received connections if not already connected to them
-        for uri in to_connect_list {
-            trace!(
-                "(TransportMemory) {} <- {}",
-                uri,
-                self.maybe_my_uri.clone().unwrap()
-            );
-            if let Ok(id) = self.connect(&uri) {
-                // Note: Add IncomingConnectionEstablished events at start of outbox
-                // so they can be processed first.
-                outbox.insert(0, TransportEvent::IncomingConnectionEstablished(id));
+        for (uri, in_cid) in to_connect_list {
+            trace!("(TransportMemory) {} <- {:?}", uri, self.maybe_my_uri);
+            let out_cid = self.connect(&uri)?;
+            self.inbound_connection_map
+                .insert(in_cid.clone(), out_cid.clone());
+            // Note: Add IncomingConnectionEstablished events at start of outbox
+            // so they can be processed first.
+            outbox.insert(0, TransportEvent::IncomingConnectionEstablished(out_cid));
+        }
+        // process other messages
+        for event in output {
+            match event {
+                TransportEvent::ConnectionClosed(in_cid) => {
+                    // convert inbound connectionId to outbound connectionId.
+                    // let out_cid = self.inbound_connection_map.get(&in_cid).expect("Should have outbound at this stage");
+                    let out_cid = self
+                        .inbound_connection_map
+                        .remove(&in_cid)
+                        .expect("Should have outbound at this stage");
+                    // close will fail as other side isn't there anymore
+                    let _ = self.close(&out_cid);
+                    outbox.push(TransportEvent::ConnectionClosed(out_cid.to_string()));
+                }
+                TransportEvent::ReceivedData(in_cid, data) => {
+                    // convert inbound connectionId to outbound connectionId.
+                    let out_cid = self
+                        .inbound_connection_map
+                        .get(&in_cid)
+                        .expect("Should have outbound at this stage");
+                    outbox.push(TransportEvent::ReceivedData(
+                        out_cid.to_string(),
+                        data.clone(),
+                    ));
+                }
+                _ => unreachable!(),
```

Comment:
It's a matter of preference, but by removing this `unreachable!()` statement, and so removing the default match pattern, we can catch some missing implementation of possible future additional enum variant. What do you think?

Reply:
There is no possiblity of missing future additional enum variant because it would have to be implemented by the MemoryServer first. Here we are only reacting to whatever the MemoryServer can send us. Will add a comment to clarify.

Reply:
I meant at compile time, the Rust compiler can tell you that there is a missing branch. It's useful during refactoring. In the end it's just a nice to have, no biggie :)